### ISSUE TITLE
Modularize Classes more

### ISSUE THREAD

**Issue by M-106** [xXAI-botXx/ANN-106#27]
To adapt your code to support various types of neural networks, such as convolutional neural networks (CNNs), graph neural networks (GNNs), and other architectures, you need to modularize the network and layer designs further. Here's a breakdown of what needs to be done:

---

1. Layer Abstraction

You should create a base class Layer that defines the interface for different types of layers. Then, subclass it for specific layer types like fully connected (dense), convolutional, or pooling layers.

Example:

class Layer:
    def forward(self, X):
        raise NotImplementedError("Forward method must be implemented in the subclass.")

    def backward(self, X, grad_output):
        raise NotImplementedError("Backward method must be implemented in the subclass.")

Fully Connected Layer:

class DenseLayer(Layer):
    def __init__(self, input_size, output_size, activation_func):
        self.weights = np.random.normal(0, 0.01, (output_size, input_size))
        self.bias = np.random.normal(0, 0.01, (output_size,))
        self.activation_func = activation_func if activation_func else lambda x: x

    def forward(self, X):
        z = np.dot(self.weights, X) + self.bias
        return self.activation_func(z)

    def backward(self, X, grad_output):
        # Implement backward logic (e.g., compute gradients)
        pass

Convolutional Layer:

class ConvLayer(Layer):
    def __init__(self, input_channels, output_channels, kernel_size, stride=1, padding=0):
        self.kernel = np.random.normal(0, 0.01, (output_channels, input_channels, kernel_size, kernel_size))
        self.stride = stride
        self.padding = padding

    def forward(self, X):
        # Implement convolution operation
        pass

    def backward(self, X, grad_output):
        # Implement backward logic
        pass

---

2. Graph Neural Network Layers

For GNNs, you need layers like graph convolutional layers or graph attention layers. These involve operations on adjacency matrices or edge lists.

Example:

class GraphConvLayer(Layer):
    def __init__(self, input_size, output_size):
        self.weights = np.random.normal(0, 0.01, (input_size, output_size))
        self.bias = np.random.normal(0, 0.01, (output_size,))

    def forward(self, X, adjacency_matrix):
        # X: Node features, adjacency_matrix: Connectivity info
        Z = adjacency_matrix @ X @ self.weights + self.bias
        return Z

---

3. Add Network Flexibility

Instead of hardcoding add_layer to use DenseLayer, allow it to accept any layer type.

Updated add_layer:

def add_layer(self, layer):
    """
    Adds a layer to the network.

    :param layer: A Layer object (DenseLayer, ConvLayer, etc.).
    :type layer: Layer
    """
    self.layers.append(layer)

Now, you can use:

ann = ArtificialNeuralNetwork()
ann.add_layer(DenseLayer(128, 64, activation_func=np.tanh))
ann.add_layer(ConvLayer(3, 16, kernel_size=3))

---

4. Custom Loss Functions

For CNNs or GNNs, you might use loss functions tailored for the task (e.g., cross-entropy for classification, mean squared error for regression).

---

5. Modular forward Pass

Update the forward method to handle layers dynamically:

def forward(self, X, **kwargs):
    for layer in self.layers:
        if isinstance(layer, GraphConvLayer):
            X = layer.forward(X, kwargs.get('adjacency_matrix'))
        else:
            X = layer.forward(X)
    return X

---

6. Support Different Layer Types in Training

Ensure that the train method dynamically interacts with all layer types. For example:

Pass graph-related inputs (like adjacency matrices) for GNN layers.

Handle feature maps for CNN layers.

---

Final Considerations

Batch Normalization: Add batch normalization layers to improve training.

Dropout: Implement dropout layers to prevent overfitting.

Pooling: Add pooling layers for dimensionality reduction in CNNs.

This modular design will allow you to extend your framework to build not just fully connected networks, but also CNNs, RNNs, GNNs, and more.
